archive_converter.py

```diff
@@ -13,8 +13,57 @@
 # limitations under the License.
 
 import email
+import os
 from setup_gmail import Message
 
+class ArchiveMessageIndex(object):
+    def __init__(self):
+        self._lookup_index = {} # Maps MessageKey to Thread
+        self._messages_seen = {} # Maps message.id to message
+
+    def find_thread(self, key: str) -> Message:
+        return self._lookup_index[key]
+
+    def update(self):
+        fully_updated = False
+        new_messages = []
+
+        for filename in os.listdir('test_data'):
+            if filename.endswith(".txt"):
+                email = generate_email_from_file('test_data/' + filename)
+                if(email.id not in self._messages_seen):
+                    new_messages.append(email)
+        self._add_messages_to_index(new_messages)
+
+    def _add_messages_to_index(self, new_messages):
+        need_parent = []
+        for message in new_messages:
+            self._messages_seen[message.id] = message
+
+            if message.in_reply_to is None:
+                key = message.get_key()
+                print(key)
+                if key:
+                    self._lookup_index[key] = message
+                else:
+                    # If message doesn't have the LKML style [PATCH v4 00/11],
+                    # we don't need to put it in the index.
+                    continue
+            else:
+                need_parent.append(message)
+        for message in need_parent:
+            if message.in_reply_to not in self._messages_seen:
+                # If we don't have the parent message, then we can't reconstruct
+                # the patchset for this group of emails anyway, so just skip.
+                continue
+            parent = self._messages_seen[message.in_reply_to]
+            parent.children.append(message)
+            
+def find_thread(key: str) -> Message:
```

Comment:
I assume this function is more for convenience and that the users of this module are going to make index objects and update them manually?
(rather than calling this func and recreating the index every time)

If so, do you want to move it into the test file as a helper there?

Reply:
yes! also, finding specific threads within the index will probably become deprecated once the application is running on a server. I'll move it to the test file!

Reply:
Actually, I've thought about it a bit more. Does the location matter? The setup_gmail version of find_thread was defined in the exact same way. It just feels wrong to import find_thread from a test file, because gerrit.py's main function still uses find_thread.

Reply:
Agreed. You should not import form a test file. Nevertheless, feel free to pull non-Gmail specific stuff out of the setup_gmail file; a lot of stuff was put there that probably doesn't belong there.